### ISSUE TITLE
Tmux-like sessions for multiple servers with one ui

### ISSUE THREAD

**Issue by asmodeus812** [neovide/neovide#2174]
**Is your feature request related to a problem? Please describe.**
This is something i have been thinking about, but not entirely sure if it is possible at the moment yet. In very simple terms, have multiple neovim servers/instances connected to the same neovide ui, in a very similar fashion to tmux sessions. The use case is pretty much the same as in tmux, being able to connect to multiple neovim instances running inside dev containers (for example).  The added benefit is that we might be able to share buffer information across the sessions, or anything else for that matter, since neovide has all the context/information it needs from the running servers it has been attached to.

This would require us to introduce a concept of tmux-like sessions or something along those lines which does not really exist in neovim itself, as a concept.

**Describe the solution you'd like**
A neovide api or user command, which would allow us to connect to or even start a new instance of neovim on the remote machine - be it a container, server etc. This would somewhat extend the current behavior, to something more general, covering a wider use case. 

**Describe alternatives you've considered**
Tmux. The concept/idea is pretty much lifted off of tmux sessions.

**Additional context**
None

**Comment by IhsenBen** [comment]
I second this, at least make it possible to launch it from a Tmux session.
I often find myself going back to Alacrity just because of  Tmux.

**Comment by honeyspoon** [comment]
It would be nice if neovide would be the terminal emulator replacing Alacrity
very out of scope for the project but it would solve the problem of being way faster to navigate in tmux with nvim vs neovide

**Comment by brentd** [comment]
This would be the killer feature of neovide for me. In kitty I use tmux with [sesh](https://github.com/joshmedeski/sesh) to jump between tmux sessions of my projects (where neovim is almost always running in at least one pane.) Sesh basically combines fzf and zoxide to give you directory-based tmux sessions, probably one of many utils that do something similar.

But I don't really love tmux like others do. I'd happily do 99% of my terminal usage within neovide and neovim's terminals if it provided tmux-like sessions which remain running in the background even when switched to another session, unlike neovim's built-in sessions.

**Comment by fredizzimo** [comment]
I think we will be able to add support for this, in a not too distant future. We actually have a couple of different options

Although this initially focuses on multi-window support https://github.com/neovide/neovide/pull/2872. It also refactors the internals so that it's possible to run multiple instances of Nvim simultaneously. So, it will be relatively easy to support switching the instance within a single window.

Another approach, that might be possible with Neovim 0.11 is the built in support https://github.com/neovim/neovim/issues/5035. Even if that is imolemented, it still need some support in Neovide to be able to connect to another instance without quitting. It also needs at least these bugs fixed, so that reconnecting works
* https://github.com/neovim/neovim/issues/30581
* https://github.com/neovim/neovim/issues/30582
* https://github.com/neovim/neovim/issues/30583
* https://github.com/neovim/neovim/issues/30586

Personally, I use two different approaches to work with multiple projects in one instance within Neovim, which works quite well without any Neovide support. At least for me.

Depending on if the work is temporary or more long lived
1. Open different projects in different neovim tabs. And making sure that [tcd](https://neovim.io/doc/user/editing.html#%3Atcd) is set inside the tab rather than the global working directory.
2. Use [telescope-project](https://github.com/nvim-telescope/telescope-project.nvim) to quickly switch between project, that's also how I usually set the initial project in a tab.

Furtermore, I use a float terminal plugin, and have it setup so that it has separate terminals per `tcd`, Actually a few, for example (m)isc, (g)it, (b)uild, which I can quickly bring up with a shortcut.

**Comment by asmodeus812** [comment]
> * Open different projects in different neovim tabs. And making sure that [tcd](https://neovim.io/doc/user/editing.html#%3Atcd) is set inside the tab rather than the global working directory.
> * Use [telescope-project](https://github.com/nvim-telescope/telescope-project.nvim) to quickly switch between project, that's also how I usually set the initial project in a tab.

Well first option is what i use most of the time that works, however that will not work if the projects are in containers. Which is what the original issue was posted about.
The second option is related to the first one i guess same.

Generally What i was looking for is neovide be able to connect to multiple neovim servers running on any machine, container - remote or local. Be able to share the buffer contents, problem here is that each neovim instance is having its own global and local options and such, so i would imagine that is harder to achieve than it sounds. 

That is why i suggested a tmux like session approaach, so neovide will have an api to allo one to switch between the active / currently connected ui sessions, BUT, unlike tmux, since neovide completely understands neovim's state, i would like to have neovide be able to share some of the resources between the remote nvim server sessions, for example, be able to view a buffer/file from another session in your curent session, maybe edit it too, copy from it to local registers etc. That does seem like a complex issue for sure.

**Comment by fredizzimo** [comment]
> BUT, unlike tmux, since neovide completely understands neovim's state, i would like to have neovide be able to share some of the resources between the remote nvim server sessions, for example, be able to view a buffer/file from another session in your curent session, maybe edit it too, copy from it to local registers etc. That does seem like a complex issue for sure.

Unfortunately that's not the case, neovide only renders whatever neovim tells it to. That type of multiplexing would be better implemented by Neovim itself or even as a lua plugin. It would be copletely possible for a lua plugin to connect to anothet lnstance and fetch some data from it. Or at least I think so.

**Comment by asmodeus812** [comment]
I see, then i was thinking, a ui layer similar to the tmux sessions/windws. Similarly to how you create a new window or session in tmux and ssh into a remote machine. A command or api to connect to a remote nvim server at a given address - NewSession {remote-nvim-server-instance}. And an api to switch between the connected sessions, for starters no shared resources are needed i think. That would simulate the tmux experience in a way. The benefit here is to avoid having multiple neovide app instances opened each attached to a different host. I know that we might be able to replicate this behavior with some window managers which can have a tabbed

**Comment by jamylak** [comment]
https://github.com/neovim/neovim/issues/23093#issuecomment-2648884489
https://github.com/neovim/neovim/pull/30319

Is there a way the new features eg. `:detach` could be used to switch to a different sessions?

**Comment by fredizzimo** [comment]
Yes, properly supporting that is also on the roadmap. But currently there are a few bugs in Neovim that prevents a multigrid UI from re-attaching to a session. So, those needs to be fixed first.

**Comment by psudo123** [comment]
I think this should be a much higher priority. It would make it so you can use Neovide basically as an entire terminal emulator and tmux replacement.
